### PR TITLE
HTTP check updates

### PR DESCRIPTION
**Pull request by brendangibat**
Adds in check-port, check-http-cors, check-head-redirect, adds in some extra checks against check-http, modifies check-last-modified to be able to be able to follow a given amount of redirects, adds in a common class.

### CODE REVIEW

**Comment by analytically** [comment]
Awesome, thanks. Would you be able to write some specs? Also can you explain the AWS link?

**Comment by brendangibat** [comment]
Yep, sure thing!

**Comment by sstarcher** [comment]
@brendangibat any updates?

**Comment by brendangibat** [comment]
Sorry, got busy and dropped off a few of the sensu PRs. I'll be updating this weekend for a few of them.

**Comment by analytically** [comment]
Would love to get this merged!

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,9 @@
+root = true
```
Comment (analytically) [inline .editorconfig]: Please delete this file, this is local to your environment.

File: bin/check-head-redirect.rb [hunk 2]
```diff
@@ -0,0 +1,150 @@
+#! /usr/bin/env ruby
```
Comment (analytically) [inline bin/check-head-redirect.rb]: no space here

File: bin/check-head-redirect.rb [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+  option :aws_secret_access_key,
+         short:       '-k AWS_SECRET_KEY',
+         long:        '--aws-secret-access-key AWS_SECRET_KEY',
+         description: "AWS Secret Access Key. Either set ENV['AWS_SECRET_KEY'] or provide it as an option",
+         default:     ENV['AWS_SECRET_KEY']
+
+  option :aws_region,
+         short:       '-r AWS_REGION',
+         long:        '--aws-region REGION',
+         description: 'AWS Region (defaults to us-east-1).',
```
Comment (analytically) [inline bin/check-head-redirect.rb]: Remove the dot here

File: lib/sensu-plugins-http/common.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
+
+module Common
+  def initialize
+    super()
+    aws_config
+  end
+
+  def aws_config
+    Aws.config.update(
```
Comment (analytically) [inline lib/sensu-plugins-http/common.rb]: Mind explaining why AWS stuff has to do with HTTP checks?

File: bin/check-port.rb [hunk 5]
```diff
@@ -0,0 +1,83 @@
+#! /usr/bin/env ruby
```
Comment (analytically) [inline bin/check-port.rb]: No space here

File: lib/sensu-plugins-http/common.rb (hunk 4 above)
Comment (brendangibat) [inline lib/sensu-plugins-http/common.rb]: We needed to pull configuration from an encrypted s3 bucket

Comment (analytically) [inline lib/sensu-plugins-http/common.rb]: That's quite a unique usecase. Mind explaining how this works a bit more in detail, and add a section to the README.md.

File: bin/check-head-redirect.rb [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+require 'aws-sdk-core'
+require 'json'
+require 'sensu-plugins-http'
+
+#
+# Checks that redirection links can be followed in a set number of requests.
+#
+class CheckLastModified < Sensu::Plugin::Check::CLI
+  include Common
+  option :aws_access_key,
```
Comment (analytically) [inline bin/check-head-redirect.rb]: aws_access_key_id please

File: bin/check-head-redirect.rb [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+         short:       '-a AWS_ACCESS_KEY',
+         long:        '--aws-access-key AWS_ACCESS_KEY',
```
Comment (analytically) [inline bin/check-head-redirect.rb]: --aws-access-key-id AWS_ACCESS_KEY_ID

File: lib/sensu-plugins-http/common.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+      credentials: Aws::Credentials.new(config[:aws_access_key], config[:aws_secret_access_key])
+    ) if config[:aws_access_key] && config[:aws_secret_access_key]
+
+    Aws.config.update(
+      region: config[:aws_region]
```
Comment (analytically) [inline lib/sensu-plugins-http/common.rb]: and environment var AWS_DEFAULT_REGION

File: bin/check-head-redirect.rb [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+         default:     ENV['AWS_SECRET_KEY']
+
+  option :aws_region,
+         short:       '-r AWS_REGION',
+         long:        '--aws-region REGION',
+         description: 'AWS Region (defaults to us-east-1)',
+         default:     'us-east-1'
+
+  option :s3_config_bucket,
+         short:       '-s S3_CONFIG_FILE',
```
Comment (analytically) [inline bin/check-head-redirect.rb]: bucket

File: bin/check-head-redirect.rb [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+         long:        '--s3-config-file S3_CONFIG_FILE',
+         description: 'S3 config bucket'
+
+  option :s3_config_key,
+         short:       '-k S3_CONFIG_KEY',
+         long:        '--s3-config-KEY S3_CONFIG_KEY',
```
Comment (analytically) [inline bin/check-head-redirect.rb]: --s3-config-key

File: bin/check-head-redirect.rb [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+require 'net/https'
+require 'time'
+require 'aws-sdk-core'
+require 'json'
+require 'sensu-plugins-http'
+
+#
+# Checks that redirection links can be followed in a set number of requests.
+#
+class CheckLastModified < Sensu::Plugin::Check::CLI
```
Comment (analytically) [inline bin/check-head-redirect.rb]: CheckLastModified??

File: lib/sensu-plugins-http/common.rb (hunk 4 above)
Comment (brendangibat) [inline lib/sensu-plugins-http/common.rb]: it pulls a json config file from an s3 bucket and merges the configuration in to the current checks config object. It's more generic than just password or username needed to be used, so if an endpoint can't be exposed either then that can be pushed in there as well.

Comment (brendangibat) [inline lib/sensu-plugins-http/common.rb]: I was using the same approach on another sensu check so I've just been making that common function generic if I need to pull secure configuration from an AWS bucket.

Comment (brendangibat) [inline lib/sensu-plugins-http/common.rb]: I changed it to not use aws if the bucket isn't configured

File: bin/check-port.rb [hunk 12]
```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env ruby
+#
+#  encoding: UTF-8
+#   check-port
```
Comment (eheydrick) [inline bin/check-port.rb]: why is check-port being added here? It's a duplicate of check-ports from sensu-plugins-network-checks.

Comment (analytically) [inline bin/check-port.rb]: Good point.

**Comment by sstarcher** [comment]
Why include the AWS S3 bucket configuration directly into the plugin?  It would be better to pull that configuration file in whatever is writing your check configs to disk instead of having the plugin know about AWS, as this could extend to any and every secret configuration request.

**Comment by analytically** [comment]
Bump.

**Comment by jkernech** [comment]
+1

File: bin/check-port.rb (hunk 12 above)
Comment (brendangibat) [inline bin/check-port.rb]: Removed.

**Comment by sstarcher** [comment]
Fails CI @brendangibat

**Comment by mattyjones** [comment]
@analytically @sstarcher I will merge this manually if we are good. This bits are fairly useful and worth the effort I feel.

**Comment by sstarcher** [comment]
@mattyjones I like everything except for the aws specific bits

**Comment by mattyjones** [comment]
@brendangibat  Any thoughts on this?

**Comment by sstarcher** [comment]
@mattyjones what's your opinion on the addition of the s3 stuff to http?

**Comment by sstarcher** [comment]
@mattyjones  @analytically opinions on this?

**Comment by analytically** [comment]
I don't have any strong opinion on the S3 stuff, as long as it doesn't impact existing users, I'm cool with it.

**Comment by sstarcher** [comment]
I have local changes for this repo rebased off of master and a few fixes to @brendangibat code. 
If everyone is fine with the s3 stuff I'll leave it or I can pull it.

**Comment by sstarcher** [comment]
merged

**Pull request closed without merging**